Add Chef Automate support to `inspec compliance login`

jerryaldrichiii opened the pull request:
This provides a single interface for logging into either Chef Automate or Chef Compliance servers. Server type is evaluated at run time via HTTP responses from designated endpoints.

This also moves the login logic from `Compliance::ComplianceCLI` to a separate set of modules in `Compliance::API`. Thus removing logic from Thor and allowing for more in depth Unit testing.

This also adds deprecation warnings for `login_automate` and `--usertoken`.

This fixes #1795.

Signed-off-by: Jerry Aldrich <jerryaldrichiii@gmail.com>

jerryaldrichiii committed:
Merge `login` and `login_automate` commands

This provides a single interface for logging into either Chef Automate
or Chef Compliance servers. Server type is evaluated at run time via
HTTP responses from designated endpoints.

This also moves the login logic from `Compliance::ComplianceCLI` to a
separate set of modules in `Compliance::API`. This removes logic from
Thor and allows for more in depth Unit testing.

Signed-off-by: Jerry Aldrich <jerryaldrichiii@gmail.com>

adamleff commented on lib/bundles/inspec-compliance/api.rb:
```diff
@@ -5,13 +5,18 @@
   # API Implementation does not hold any state by itself,
   # everything will be stored in local Configuration store
   class API # rubocop:disable Metrics/ClassLength
+
```
nit: extra newline at class beginning

jerryaldrichiii replied:
Pick all the nits! Fixed.

adamleff commented on lib/bundles/inspec-compliance/api.rb:
```diff
@@ -238,5 +243,11 @@ def self.server_version_from_config(config)
     end
+
+    def self.determine_server_type(url, insecure)
+      return 'automate' if Compliance::HTTP.get(url + '/compliance/version', nil, insecure).code == '401'
```
It's more idiomatic to return symbols for something like this rather than strings.

`return :automate if ...`

jerryaldrichiii replied:
I always prefer going the idiomatic route. Fixed. Thanks @adamleff!

adamleff commented on lib/bundles/inspec-compliance/api.rb:
```diff
@@ -238,5 +243,11 @@ def self.server_version_from_config(config)
+    def self.determine_server_type(url, insecure)
+      return 'automate' if Compliance::HTTP.get(url + '/compliance/version', nil, insecure).code == '401'
+      return 'compliance' if Compliance::HTTP.get(url + '/api/version', nil, insecure).code == '200'
+      raise CannotDetermineServerType
```
You should also add an exception message when raising.

`raise CannotDetermineServerType, "Unable to determine if #{url} is a Compliance or Automate server"`

jerryaldrichiii replied:
Agreed. Fixed.

adamleff commented on lib/bundles/inspec-compliance/api/login.rb:
```diff
@@ -0,0 +1,134 @@
+
+    options['url'] = options['server'] + '/compliance'
+    token = options['dctoken'] || options['token']
+    puts store_access_token(options, token)
```
I know it's a WIP, to just a reminder to remove this :)

jerryaldrichiii replied:
`store_access_token` returns a string at this time. I will look at the flow and see if we should pass this up as a return from `login`.

Will investigate.

adamleff commented on lib/bundles/inspec-compliance/api/login.rb:
```diff
@@ -0,0 +1,134 @@
+  def login(options)
+    options['server_type'] = Compliance::API.determine_server_type(options['server'], options['insecure'])
+
+    return Login::ComplianceServer.login(options) if options['server_type'] == 'compliance'
```
Will you be doing the Automate login here too? Probably worth Zooming to review the expected flow, etc. would be good when you're ready :) This mid-method return with no comments as to why is confusing me as to your intentions for this method and how Automate login will be handled.

jerryaldrichiii replied:
This redirects to the `login` method of the `ComplianceServer` module if the server type is a Compliance server. I was returning here, since this will complete the login. Initially I had it just `exit 0` at the end, but that was causing me issues with the Unit tests.

Will investigate.

clintoncwolfe replied:
I'd second @adamleff here - If you're switching between two sets of logic, and you've encapsulated one in a module, I'd expect the other logic - for Automate login - to also be encapsulated in a module (presumably Login::AutomateServer).

jerryaldrichiii replied:
Originally I thought only to encapsulate the Compliance server stuff because it would be deprecated in the future leaving only `Login`...but the future is not now. I see your point.

Fixed.

jerryaldrichiii replied:
@adamleff let me know what you think of the flow now that I've modified the namespacing.

adamleff commented on lib/bundles/inspec-compliance/api/login.rb:
```diff
@@ -0,0 +1,134 @@
+  private
+
+  def store_access_token(options, token)
+    token_info = { type: 'dctoken', msg: 'data colletor token' } if options['dctoken']
```
Will you ever not have either a dctoken or a token? What happens if token_info is not set?

I'd recommend changing it to something like:

```ruby
token_info = if options['dctoken']
               { # dctoken stuff }
             else
               { # non-dctoken stuff }
             end
```

jerryaldrichiii replied:
I agree. Fixed.

adamleff reviewed (commented):
Just flipped through this quickly, some comments and questions, probably worth a Zoom review when it's no longer WIP.

jerryaldrichiii committed:
Remove empty line below class definition

Signed-off-by: Jerry Aldrich <jerryaldrichiii@gmail.com>

jerryaldrichiii committed:
Add message to `raise CannotDetermineServerType`

Signed-off-by: Jerry Aldrich <jerryaldrichiii@gmail.com>

jerryaldrichiii committed:
Refactor `token_info` assignment

Signed-off-by: Jerry Aldrich <jerryaldrichiii@gmail.com>

clintoncwolfe commented on lib/bundles/inspec-compliance/cli.rb:
```diff
@@ -18,78 +18,32 @@ def self.subcommand_prefix
-    option :insecure, aliases: :k, type: :boolean,
-      desc: 'Explicitly allows InSpec to perform "insecure" SSL connections and transfers'
-    def login_automate(server) # rubocop:disable Metrics/AbcSize
+    def login(server) # rubocop:disable Metrics/AbcSize
```
You don't need the rubocop disable flag here anymore (when you refactored this out, the method became much simpler).    You can run a rubocop scan with 
```bundle exec rake lint```

jerryaldrichiii replied:
Great find! Fixed.

jerryaldrichiii replied:
Agreed, nice catch! Fixed.

clintoncwolfe commented on lib/bundles/inspec-compliance/api.rb:
```diff
@@ -5,13 +5,17 @@
   # API Implementation does not hold any state by itself,
   # everything will be stored in local Configuration store
   class API # rubocop:disable Metrics/ClassLength
+    extend Login
```
This may be a matter of taste, but I'd usually have the module name here more deeply namespaced - like Compliance::API::Login .  I don't know that we'd ever (within InSpec) have a name collision with a module named Login, but being specific avoids that possibility.

jerryaldrichiii replied:
Agreed! Fixed.

jerryaldrichiii committed:
Remove unnecessary rubocop disable

Signed-off-by: Jerry Aldrich <jerryaldrichiii@gmail.com>

jerryaldrichiii committed:
Modify `Login` module namespacing

Signed-off-by: Jerry Aldrich <jerryaldrichiii@gmail.com>

jerryaldrichiii committed:
Remove mentions of login_automate and --usertoken

Signed-off-by: Jerry Aldrich <jerryaldrichiii@gmail.com>

jerryaldrichiii committed:
Modify `determine_server_type` to return a symbol

Signed-off-by: Jerry Aldrich <jerryaldrichiii@gmail.com>

jerryaldrichiii committed:
Add support for `login_automate` and `--usertoken`

Signed-off-by: Jerry Aldrich <jerryaldrichiii@gmail.com>

jerryaldrichiii committed:
Fix encoding typo

Signed-off-by: Jerry Aldrich <jerryaldrichiii@gmail.com>

adamleff commented on lib/bundles/inspec-compliance/api/login.rb:
```diff
@@ -0,0 +1,145 @@
+  class API
+    module Login
+      def login(options)
+        options['server_type'] = Compliance::API.determine_server_type(options['server'], options['insecure']).to_s
```
Coercing to a string here seems... weird? We don't need it to be a string here or most of these methods -- only when we store the config to disk. Can we move the coercion there?

adamleff commented on lib/bundles/inspec-compliance/api.rb:
```diff
@@ -238,5 +242,11 @@ def self.server_version_from_config(config)
+    def self.determine_server_type(url, insecure)
+      return :automate if Compliance::HTTP.get(url + '/compliance/version', nil, insecure).code == '401'
+      return :compliance if Compliance::HTTP.get(url + '/api/version', nil, insecure).code == '200'
+      raise CannotDetermineServerType, "Unable to determine if #{url} is a Chef Automate or Chef Compliance server"
```
Just tossing out a thought here... I prefer to have exceptions raised at the point of the implementation where "it matters" (for lack of a better phrase). I'd almost prefer to see this method return `nil` if it can't figure out the server type, and let the caller determine how to handle. That way other methods can use this if needed without having to worry about the stack unwinding on them.

```ruby
def self.determine_server_type(url, insecure)
  if Compliance::HTTP.get(url + '/compliance/version', nil, insecure).code == '401'
    :automate
  elsif Compliance::HTTP.get(url + '/api/version', nil, insecure).code == '200'
    :compliance
  end
end
```

adamleff commented on lib/bundles/inspec-compliance/api/login.rb:
```diff
@@ -0,0 +1,145 @@
+      def login(options)
+        options['server_type'] = Compliance::API.determine_server_type(options['server'], options['insecure']).to_s
+
+        return Login::ComplianceServer.login(options) if options['server_type'] == 'compliance'
```
Here is where I'd raise an exception if we don't know the server type since this is actually where an error should be raised to the user:

```ruby
if options['server_type'] == :compliance
  Login::ComplianceServer.login(options)
elsif options['server_type'] == :automate
  Login::AutomateServer.login(options)
else
  raise CannotDetermineServerType, # .....
end
```

adamleff commented on lib/bundles/inspec-compliance/api/login.rb:
```diff
@@ -0,0 +1,145 @@
+
+          options['url'] = options['server'] + '/compliance'
+          token = options['dctoken'] || options['token']
+          puts store_access_token(options, token)
```
I realize I'm being a bit pedantic here, but it feels strange to me to have a method called `store_access_token` return a string intended to be user-facing messaging.

I'd rather see `store_access_token` return the `Compliance::Configuration` object it generated to successfully write the config to disk, and then a method that takes that config object and formats a message to the user. It will keep the concerns of the methods succinct and clear. This honestly just reads like a debug message to me. :)

```ruby
config = store_access_token(options, token)

# print out a success message to the user
puts format_status_message(config)
```

adamleff commented on lib/bundles/inspec-compliance/api/login.rb:
```diff
@@ -0,0 +1,145 @@
+
+          raise msg unless success
+
+          puts msg
```
Same feedback as above - I bet there's a good opportunity for reuse here, actually.

adamleff commented on lib/bundles/inspec-compliance/cli.rb:
```diff
@@ -18,78 +18,67 @@ def self.subcommand_prefix
-        msg = login_automate_config(url, options['user'], options['dctoken'], options['usertoken'], options['ent'], options['insecure'])
-        puts '', msg
-        exit 0
+      if options['usertoken']
```
Despite what was in the issue that preceded this PR, we probably don't need this deprecation here since the whole command is deprecated. As long as the `inspec compliance login` command doesn't accept a `--usertoken` flag, we're covered.

adamleff reviewed (changes requested):
This is in great shape, @jerryaldrichiii. I'd like to see a bit of cleanup around where we raise exceptions and keeping method purposes focused. Let me know if any of my feedback is unclear.

jerryaldrichiii committed:
Address PR feedback

This does the following:
  - Moves `CannotDetermineServerType` error to `.login`
  - Changes methods that store configuration to return the configuration
  - Moves user output to one location in `.login`
  - Makes other small improvements

Signed-off-by: Jerry Aldrich <jerryaldrichiii@gmail.com>

adamleff reviewed (approved):
Lookin' good, @jerryaldrichiii - thanks!

chris-rock commented on lib/bundles/inspec-compliance/api.rb:
```diff
@@ -5,13 +5,16 @@
 require 'net/http'
 require 'uri'
 
+require_relative 'api/login'
```
should we move this file to the api directory too? Feels strange to have a `api.rb` and `api` directory. What do you think?

adamleff replied:
That's actually pretty standard in Ruby.

jerryaldrichiii replied:
Yeah, it's common in Ruby. I'm fine with it if you are @chris-rock.

chris-rock commented on lib/bundles/inspec-compliance/api.rb:
```diff
@@ -238,5 +241,13 @@ def self.server_version_from_config(config)
     end
+
+    def self.determine_server_type(url, insecure)
+      if Compliance::HTTP.get(url + '/compliance/version', nil, insecure).code == '401'
```
Could this also return 200 if the user is already logged in?

jerryaldrichiii replied:
I'm not passing any headers to the `GET` request so it should always act as if the user is not logged in.

chris-rock commented on lib/bundles/inspec-compliance/api/login.rb:
```diff
@@ -0,0 +1,150 @@
+
+        case options['server_type']
+        when :automate
+          config = Login::AutomateServer.login(options)
```
Nice separation!

jerryaldrichiii replied:
Thanks! Took some work that's for sure :sweat_smile:

chris-rock commented on lib/bundles/inspec-compliance/cli.rb:
```diff
@@ -18,78 +18,64 @@ def self.subcommand_prefix
-      end
-
-      puts '', msg
+      Compliance::API.login(options)
```
Nice clean-up! 👍

jerryaldrichiii replied:
Thanks! :smiley:

chris-rock reviewed (commented):
@jerryaldrichiii Awesome work. I just have some mini-questions before we can go forward

chris-rock reviewed (approved):
Works like a charm!
